File: frontend_old/websites/src/website/packages/components/src/api/moderation/category/ListTtsCategoriesForModeration.ts
import { ApiConfig } from "@storyteller/components";

export enum ListTtsCategoriesTriState {
  Include,
  Exclude,
  Only,
}

export interface ListTtsCategoriesForModerationSuccessResponse {
  success: boolean,
  categories: Array<ModerationTtsCategory>
}

export interface ModerationTtsCategory {
  category_token: string,
  model_type: string,
  maybe_super_category_token?: string,

  can_directly_have_models: boolean,
  can_have_subcategories: boolean,
  can_only_mods_apply: boolean,

  name: string,
  maybe_dropdown_name?: string,

  creator_user_token?: string,
  creator_username?: string,
  creator_display_name?: string,
  creator_gravatar_hash?: string,

  is_mod_approved?: boolean,
  maybe_mod_comments?: string,

  created_at: Date,
  updated_at: Date,
  deleted_at?: Date,
}

export interface ListTtsCategoriesForModerationErrorResponse {
  success: boolean,
}

type ListTtsCategoriesForModerationResponse = ListTtsCategoriesForModerationSuccessResponse | ListTtsCategoriesForModerationErrorResponse;

export function ListTtsCategoriesForModerationIsOk(response: ListTtsCategoriesForModerationResponse): response is ListTtsCategoriesForModerationSuccessResponse {
  return response?.success === true;
}

export function ListTtsCategoriesForModerationIsError(response: ListTtsCategoriesForModerationResponse): response is ListTtsCategoriesForModerationErrorResponse {
  return response?.success === false;
}

export async function ListTtsCategoriesForModeration(deleted: ListTtsCategoriesTriState, unapproved: ListTtsCategoriesTriState) : Promise<ListTtsCategoriesForModerationResponse> 
{
  let endpoint = new ApiConfig().getModerationTtsCategoryList();

  let deletedState = TriStateToString(deleted);
  let unapprovedState = TriStateToString(unapproved);

  endpoint = `${endpoint}?deleted=${deletedState}&unapproved=${unapprovedState}`;
  
  return await fetch(endpoint, {
    method: 'GET',
    headers: {
      'Accept': 'application/json',
    },
    credentials: 'include',
  })
  .then(res => res.json())
  .then(res => {
    if (!res) {
      return { success : false }; // TODO: This loses error semantics and is deprecated
    }

    if (res && 'success' in res) {
      return res;
    } else {
      return { success : false };
    }
  })
  .catch(e => {
    return { success : false };
  });
}

function TriStateToString(state: ListTtsCategoriesTriState) : string {
  switch (state) {
    case ListTtsCategoriesTriState.Include:
      return 'include';
    case ListTtsCategoriesTriState.Exclude:
      return 'exclude';
    case ListTtsCategoriesTriState.Only:
      return 'only';
  }
}

File: crates/service/web/storyteller_web/src/http_server/deprecated_endpoints/moderation/categories/list_tts_categories_for_moderation.rs
use std::fmt;
use std::sync::Arc;

use actix_web::error::ResponseError;
use actix_web::http::StatusCode;
use actix_web::{web, HttpRequest, HttpResponse};
use chrono::{DateTime, Utc};
use log::warn;

use http_server_common::response::serialize_as_json_error::serialize_as_json_error;
use mysql_queries::queries::model_categories::list_categories_query_builder::ListCategoriesQueryBuilder;

use crate::state::server_state::ServerState;

// =============== Request ===============

#[derive(Deserialize, Copy, Clone, Debug)]
#[serde(rename_all = "snake_case")]
pub enum ViewState {
  /// 'exclude'; keep the objects out
  Exclude,
  /// 'include'; include the objects
  Include,
  /// 'only'; limit to just the objects
  Only,
}

// Only show approved&non-deleted , only show un-approved, only show deleted
#[derive(Deserialize)]
pub struct QueryParams {
  deleted: Option<ViewState>,
  unapproved: Option<ViewState>,
}

// =============== Success Response ===============

#[derive(Serialize)]
pub struct ListTtsCategoriesForModerationResponse {
  pub success: bool,
  pub categories: Vec<CategoryForModeration>,
}

#[derive(Serialize)]
pub struct CategoryForModeration {
  pub category_token: String,

  pub model_type: String, // TODO: ENUM

  pub maybe_super_category_token: Option<String>,

  pub can_directly_have_models: bool,
  pub can_have_subcategories: bool,
  pub can_only_mods_apply: bool,

  pub name: String,
  pub maybe_dropdown_name: Option<String>,

  pub creator_user_token: Option<String>,
  pub creator_username: Option<String>,
  pub creator_display_name: Option<String>,
  pub creator_gravatar_hash: Option<String>,

  // Moderator fields
  pub is_mod_approved: Option<bool>,
  pub maybe_mod_comments: Option<String>,

  pub created_at: DateTime<Utc>,
  pub updated_at: DateTime<Utc>,
  pub deleted_at: Option<DateTime<Utc>>,
}

// =============== Error Response ===============

#[derive(Debug, Serialize)]
pub enum ListTtsCategoriesForModerationError {
  NotAuthorized,
  ServerError,
}

impl ResponseError for ListTtsCategoriesForModerationError {
  fn status_code(&self) -> StatusCode {
    match *self {
      ListTtsCategoriesForModerationError::NotAuthorized => StatusCode::UNAUTHORIZED,
      ListTtsCategoriesForModerationError::ServerError => StatusCode::INTERNAL_SERVER_ERROR,
    }
  }

  fn error_response(&self) -> HttpResponse {
    serialize_as_json_error(self)
  }
}

// NB: Not using derive_more::Display since Clion doesn't understand it.
impl fmt::Display for ListTtsCategoriesForModerationError {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    write!(f, "{:?}", self)
  }
}

// =============== Handler ===============

pub async fn list_tts_categories_for_moderation_handler(
  http_request: HttpRequest,
  query: web::Query<QueryParams>,
  server_state: web::Data<Arc<ServerState>>) -> Result<HttpResponse, ListTtsCategoriesForModerationError>
{
  let maybe_user_session = server_state
      .session_checker
      .maybe_get_user_session(&http_request, &server_state.mysql_pool)
      .await
      .map_err(|e| {
        warn!("Session checker error: {:?}", e);
        ListTtsCategoriesForModerationError::ServerError
      })?;

  let user_session = match maybe_user_session {
    Some(session) => session,
    None => {
      warn!("not logged in");
      return Err(ListTtsCategoriesForModerationError::NotAuthorized);
    }
  };

  // TODO: We don't have a permission for categories, so we use this as a proxy.
  if !user_session.can_ban_users {
    warn!("no permission to edit categories");
    return Err(ListTtsCategoriesForModerationError::NotAuthorized);
  }

  // The objects we'll show or hide
  let delete_state = query.deleted.unwrap_or(ViewState::Exclude);
  let unapproved_state = query.unapproved.unwrap_or(ViewState::Include);

  let mut query_builder = ListCategoriesQueryBuilder::new()
      .scope_model_type(Some("tts"));

  query_builder = match delete_state {
    ViewState::Exclude => query_builder.show_deleted(false),
    ViewState::Include => query_builder.show_deleted(true),
    ViewState::Only => query_builder.show_deleted(true).hide_non_deleted(true),
  };

  query_builder = match unapproved_state {
    ViewState::Exclude => query_builder.show_unapproved(false),
    ViewState::Include => query_builder.show_unapproved(true),
    ViewState::Only => query_builder.show_unapproved(true).hide_approved(true),
  };

  let query_result =
      query_builder.perform_query(&server_state.mysql_pool).await;

  let results = match query_result {
    Ok(results) => results,
    Err(err) => {
      warn!("DB error: {:?}", err);
      return Err(ListTtsCategoriesForModerationError::ServerError);
    }
  };

  warn!("Number of categories: {:?}", results.categories.len());

  let mut categories = results.categories.iter()
      .map(|c| {
        CategoryForModeration {
          category_token: c.category_token.clone(),
          model_type: c.model_type.clone(),
          maybe_super_category_token: c.maybe_super_category_token.clone(),
          can_directly_have_models: c.can_directly_have_models,
          can_have_subcategories: c.can_have_subcategories,
          can_only_mods_apply: c.can_only_mods_apply,
          name: c.name.clone(),
          maybe_dropdown_name:c.maybe_dropdown_name.clone(),
          creator_user_token: c.creator_user_token.clone(),
          creator_username: c.creator_username.clone(),
          creator_display_name: c.creator_display_name.clone(),
          creator_gravatar_hash: c.creator_gravatar_hash.clone(),
          is_mod_approved: c.is_mod_approved,
          maybe_mod_comments: c.maybe_mod_comments.clone(),
          created_at: c.created_at,
          updated_at: c.updated_at,
          deleted_at: c.deleted_at,
        }
      })
      .collect::<Vec<CategoryForModeration>>();

  // TODO: Sort by dropdown name too.
  categories.sort_by(|c1, c2| c1.name.cmp(&c2.name));

  let response = ListTtsCategoriesForModerationResponse {
    success: true,
    categories,
  };

  let body = serde_json::to_string(&response)
      .map_err(|e| ListTtsCategoriesForModerationError::ServerError)?;

  Ok(HttpResponse::Ok()
      .content_type("application/json")
      .body(body))
}
